File: sdk/python/src/fr_ir/compiler_evidence.py
"""Bound compiler observations without conflating compilation and syntax acceptance."""
from __future__ import annotations

from dataclasses import dataclass, replace
import hashlib
import json
from pathlib import Path
import re
import tempfile
from typing import Any, Mapping

from .context import ObjectStore, restore_stored_value, store_merkle_value
from .investigation import Evidence, EvidenceKind, TaskPlan, ResumedPlan
from .investigation_checks import attach_checks
from .runtime import FrClient, FrReport, FrRuntimeError, Occurrence


def _digest(value: Any, *, canonical: bool = True) -> str:
    return hashlib.sha256(json.dumps(value, sort_keys=canonical, ensure_ascii=False, separators=(",", ":")).encode()).hexdigest()


def _hash(value: Any, prefix: str = "") -> bool:
    return isinstance(value, str) and re.fullmatch(re.escape(prefix) + r"[0-9a-f]{64}", value) is not None


@dataclass(frozen=True)
class CompilerSpan:
    status: str
    primary: bool | None
    syntax: str
    occurrence: Occurrence | None
    source: Mapping[str, Any] | None

    def reveal(self, client: FrClient) -> FrReport:
        if self.occurrence is None or self.source is None:
            raise FrRuntimeError("compiler span has no source action")
        arguments = self.source.get("arguments")
        if not isinstance(arguments, list) or arguments[:2] != ["project", "show"] or any(not isinstance(x, str) for x in arguments):
            raise FrRuntimeError("expected a read-only source action")
        result = client.call(*arguments)
        if result.at("/revision") != self.occurrence.revision:
            raise FrRuntimeError("compiler source action crossed a revision")
        return result


@dataclass(frozen=True)
class CompilerDiagnostic:
    id: str
    index: int
    parent: int | None
    level: str
    code: str | None
    message: str
    spans: tuple[CompilerSpan, ...]


@dataclass(frozen=True)
class CompilerCollection:
    items: tuple[CompilerDiagnostic, ...]
    complete: bool
    disclosure_complete: bool
    pages: int
    continuation: Mapping[str, Any] | None


@dataclass(frozen=True)
class CompilerEvidence:
    report: FrReport
    items: tuple[CompilerDiagnostic, ...]
    checks: FrReport | None = None

    @property
    def complete(self) -> bool:
        return self.report.at("/complete")

    @classmethod
    def inspect(cls, client: FrClient, checks: FrReport, *, check: str, format: str = "rustc-json",
                limit: int = 8, max_bytes: int = 32768, cursor: str | None = None) -> CompilerEvidence:
        if checks.schema != "fr-checks-1" or format not in {"rustc-json", "cargo-json"}:
            raise FrRuntimeError("compiler evidence needs retained checks and an admitted format")
        value = checks.to_data()
        with tempfile.TemporaryDirectory(prefix="fr-compiler-evidence-") as directory:
            source = Path(directory) / "checks.json"
            source.write_text(json.dumps(value, ensure_ascii=False), encoding="utf-8")
            args = ["compiler-evidence", "--from", str(source), "--digest", _digest(value), "--check", check,
                    "--format", format, "--limit", str(limit), "--bytes", str(max_bytes)]
            if cursor is not None:
                args.extend(["--cursor", cursor])
            report = client.project(*args)
        return cls.from_report(report, checks=checks)

    @classmethod
    def from_report(cls, report: FrReport, *, checks: FrReport | None = None) -> CompilerEvidence:
        try:
            data = report.to_data()
            inputs, basis, revision = data["inputs"], data["basis"], data["revision"]
            if (report.schema != "fr-compiler-evidence-1" or data["mutation_authority"] is not False
                    or not _hash(revision) or basis != "frcpe1:" + _digest(inputs)
                    or inputs["revision"] != revision or inputs["analyzer"] != data["analyzer"]
                    or inputs["toolchain"] != _digest(data["scope"]["toolchain"])
                    or inputs["check"] != data["scope"]["check"] or inputs["format"] != data["scope"]["format"]
                    or inputs["format"] not in {"rustc-json", "cargo-json"}
                    or any(not _hash(inputs[key]) for key in ["configuration", "source_revision", "report", "analyzer"])):
                raise FrRuntimeError("compiler evidence has inconsistent input identities")
            if checks is not None and (checks.schema != "fr-checks-1" or inputs["report"] != _digest(checks.to_data())):
                raise FrRuntimeError("compiler evidence belongs to another retained check report")
            if checks is not None:
                retained = checks.to_data()
                selected = [row for row in retained["results"] if row["name"] == inputs["check"]]
                if (len(selected) != 1 or inputs["configuration"] != retained["basis"]
                        or inputs["source_revision"] != retained["source_revision"]
                        or data["scope"]["toolchain"] != retained["toolchain"]
                        or any(data["scope"][key] != selected[0][key] for key in ["argv", "cwd", "covers"])
                        or any(data["outcome"][key] != selected[0][key] for key in ["passed", "exit_code"])):
                    raise FrRuntimeError("compiler scope or outcome differs from retained execution")
            items = []
            for row in data["items"]:
                core = row["core"]
                index, parent = core["index"], core["parent"]
                if (row["basis"] != basis or row["id"] != "frcd1:" + _digest([basis, core])
                        or core["rule"] != "rust-compiler-evidence-1:diagnostic"
                        or core["confidence"] != "reported-by-declared-check"
                        or type(index) is not int or index < 0
                        or parent is not None and (type(parent) is not int or not 0 <= parent < index)
                        or not isinstance(core["message"], str) or len(core["message"]) > 512
                        or core["code"] is not None and not isinstance(core["code"], str)
                        or not isinstance(core["level"], str) or not isinstance(core["spans"], list) or len(core["spans"]) > 16):
                    raise FrRuntimeError("compiler diagnostic has inconsistent identity or parent")
                spans = []
                for span in core["spans"]:
                    point = None if span["occurrence"] is None else Occurrence.from_data(span["occurrence"])
                    if (span["status"] not in {"exact", "invalid", "unavailable", "expanded"}
                            or (span["status"] == "exact") != (point is not None)
                            or span["syntax"] not in {"accepted", "rejected", "unknown"}
                            or point is None and span["syntax"] != "unknown"
                            or span["primary"] is not None and type(span["primary"]) is not bool):
                        raise FrRuntimeError("compiler span has inconsistent mapping coverage")
                    if point is not None and (point.revision != revision or point.role != (
                            "compiler-primary" if span["primary"] else "compiler-secondary") or point.id != "fro1:" + _digest([
                                revision, point.path, {"start":point.location.span.start,"end":point.location.span.end},point.role], canonical=False)):
                        raise FrRuntimeError("compiler occurrence has inconsistent identity")
                    spans.append(CompilerSpan(span["status"], span["primary"], span["syntax"], point, span["source"]))
                items.append(CompilerDiagnostic(row["id"], index, parent, core["level"], core["code"], core["message"], tuple(spans)))
            page, capture, outcome = data["page"], data["capture"], data["outcome"]
            if (any(type(page[key]) is not int or page[key] < 0 for key in ["before", "returned", "remaining", "total"])
                    or page["returned"] != len(items) or page["total"] != page["before"] + len(items) + page["remaining"]
                    or (page["next"] is None) != (page["remaining"] == 0)
                    or (data["continuation"] is None) != (page["next"] is None)
                    or capture["diagnostics"] != page["total"]
                    or any(item.index != page["before"] + i for i, item in enumerate(items))):
                raise FrRuntimeError("compiler diagnostic page has inconsistent coverage")
            disclosed = page["before"] == 0 and page["remaining"] == 0
            if (any(type(value) is not bool for value in [capture["complete"], outcome["execution_complete"], outcome["passed"], data["complete"],data["disclosure_complete"]])
                    or capture["complete"] != (capture["cutoffs"] == [])
                    or data["disclosure_complete"] != disclosed
                    or data["complete"] != (capture["complete"] and outcome["execution_complete"] and disclosed)):
                raise FrRuntimeError("compiler completeness differs from capture or disclosure")
            expected = [item.id for item in items if item.level in {"error", "ice"} and any(span.syntax == "accepted" for span in item.spans)]
            if ([row["diagnostic"] for row in data["disagreements"]] != expected
                    or any(row["kind"] != "syntax-accepted-compiler-error" for row in data["disagreements"])):
                raise FrRuntimeError("compiler disagreement differs from syntax observations")
            return cls(report, tuple(items), checks)
        except (KeyError, TypeError, ValueError, AttributeError) as error:
            raise FrRuntimeError("malformed compiler evidence report") from error

    def _read(self, client: FrClient, cursor: str | None) -> CompilerEvidence:
        if self.checks is None:
            raise FrRuntimeError("retained checks are required to refresh compiler evidence")
        value = self.inspect(client, self.checks, check=self.report.at("/scope/check"), format=self.report.at("/scope/format"),
                             limit=self.report.at("/budget/limit"), max_bytes=self.report.at("/budget/response_bytes"), cursor=cursor)
        if value.report.at("/basis") != self.report.at("/basis"):
            raise FrRuntimeError("compiler evidence crossed an input identity")
        return value

    def next(self, client: FrClient) -> CompilerEvidence:
        cursor = self.report.at("/page/next")
        if cursor is None:
            raise FrRuntimeError("compiler evidence has no continuation")
        result = self._read(client, cursor)
        if result.report.at("/page/before") != self.report.at("/page/before") + len(self.items):
            raise FrRuntimeError("compiler continuation skipped diagnostics")
        return result

    def collect(self, client: FrClient, *, max_pages: int = 64) -> CompilerCollection:
        if type(max_pages) is not int or not 1 <= max_pages <= 64:
            raise FrRuntimeError("compiler collection needs a page budget between 1 and 64")
        current, items, pages = self, list(self.items), 1
        while current.report.at("/continuation") is not None and pages < max_pages:
            current = current.next(client)
            items.extend(current.items)
            pages += 1
        disclosed = self.report.at("/page/before") == 0 and current.report.at("/continuation") is None
        return CompilerCollection(tuple(items), disclosed and self.report.at("/capture/complete") and self.report.at("/outcome/execution_complete"),
                                  disclosed, pages, current.report.at("/continuation"))

    def persist(self, store: ObjectStore) -> str:
        return store_merkle_value(store, {"report":self.report.to_data(), "checks":None if self.checks is None else self.checks.to_data()}).digest

    @classmethod
    def restore(cls, store: ObjectStore, digest: str) -> CompilerEvidence:
        value = restore_stored_value(store, digest)
        return cls.from_report(FrReport(value["report"], ()), checks=None if value["checks"] is None else FrReport(value["checks"], ()))

    def attach(self, plan: TaskPlan, client: FrClient, step: str) -> ResumedPlan:
        fresh = self._read(client, None)
        if self.checks is None:
            raise FrRuntimeError("compiler attachment requires retained checks")
        resumed = attach_checks(plan, client, step, self.checks)
        observation = Evidence("compiler:" + self.report.at("/scope/check"), EvidenceKind.OBSERVATION,
            resumed.input_digests[step], fresh.complete,
            "fr-compiler-evidence-1:" + _digest(self.report.to_data()))
        steps = tuple(replace(item, evidence=tuple(e for e in item.evidence if e.id != observation.id) + (observation,))
                      if item.id == step else item for item in resumed.plan.steps)
        return replace(resumed.plan, steps=steps).resume(client)

File: src/project/compiler_evidence.rs
use super::{compiler_diagnostics::Diagnostics, hash, page, Project};
use anyhow::{ensure, Context, Result};
use clap::{Args, ValueEnum};
use serde_json::{json, Value};
use std::path::{Component, Path, PathBuf};

#[derive(Clone, Copy, ValueEnum)]
enum Format {
    RustcJson,
    CargoJson,
}

#[derive(Args)]
pub struct Options {
    #[arg(long)]
    from: PathBuf,
    #[arg(long)]
    digest: String,
    #[arg(long)]
    check: String,
    #[arg(long, value_enum, default_value = "rustc-json")]
    format: Format,
    #[arg(long, default_value_t = 8)]
    limit: usize,
    #[arg(long)]
    cursor: Option<String>,
    #[arg(long, default_value_t = 32768)]
    bytes: usize,
}

pub fn compiler_evidence_complete(
    execution: bool,
    protocol: bool,
    from_start: bool,
    no_remaining: bool,
) -> bool {
    execution && protocol && from_start && no_remaining
}

impl Project<'_> {
    pub(super) fn compiler_evidence(&self, options: &Options) -> Result<Value> {
        ensure!(
            (1..=64).contains(&options.limit),
            "diagnostic limit must be between 1 and 64."
        );
        ensure!(
            (4096..=1_048_576).contains(&options.bytes),
            "compiler response budget must be between 4096 and 1048576."
        );
        ensure!(
            std::fs::metadata(&options.from)?.len() <= 1_048_576,
            "retained check report exceeds 1 MiB."
        );
        let bytes = crate::vfs::read(&options.from)?;
        ensure!(
            bytes.len() <= 1_048_576,
            "retained check report exceeds 1 MiB."
        );
        let report: Value = serde_json::from_slice(&bytes)?;
        ensure!(
            hash(&report)? == options.digest,
            "retained check report digest mismatch."
        );
        crate::checks::evidence::validate(&self.root, &report)?;
        let check = report["results"]
            .as_array()
            .unwrap()
            .iter()
            .find(|row| row["name"] == options.check)
            .context("selected check has no retained result.")?;
        let cargo = matches!(options.format, Format::CargoJson);
        let format = if cargo { "cargo-json" } else { "rustc-json" };
        let stream = if cargo { "stdout" } else { "stderr" };
        let mut parsed = Diagnostics::read(&check[stream], cargo)?;
        let execution = check["exit_code"].is_i64()
            && check["timed_out"] == false
            && check["output_limit_exceeded"] == false
            && check["error"].is_null()
            && check["termination_error"].is_null();
        if !execution {
            parsed.cutoffs.insert("unfinished-check".into());
        }
        let errors = parsed
            .rows
            .iter()
            .filter(|row| matches!(row["level"].as_str(), Some("error" | "ice")))
            .count();
        if errors > 0 && check["passed"] == true {
            parsed.cutoffs.insert("errors-with-passing-check".into());
        }
        if parsed
            .build_success
            .is_some_and(|success| check["passed"] != success)
        {
            parsed
                .cutoffs
                .insert("cargo-check-outcome-disagreement".into());
        }
        let analyzer = hash((
            "rust-compiler-evidence-1",
            include_str!("compiler_evidence.rs"),
            include_str!("compiler_diagnostics.rs"),
        ))?;
        let inputs = json!({"revision":self.revision,"source_revision":report["source_revision"],
            "configuration":report["basis"],"toolchain":hash(&report["toolchain"])?,
            "report":options.digest,"check":options.check,"format":format,"analyzer":analyzer});
        let basis = format!("frcpe1:{}", hash(&inputs)?);
        let cwd = check["cwd"]
            .as_str()
            .context("check directory is missing.")?;
        let mut rows = Vec::new();
        let mut disagreements = Vec::new();
        for mut core in parsed.rows {
            let mut spans = Vec::new();
            for span in core["spans"].as_array().unwrap() {
                spans.push(self.compiler_span(cwd, span)?);
            }
            core["spans"] = json!(spans);
            core["rule"] = json!("rust-compiler-evidence-1:diagnostic");
            core["confidence"] = json!("reported-by-declared-check");
            let id = format!("frcd1:{}", hash((&basis, &core))?);
            if matches!(core["level"].as_str(), Some("error" | "ice"))
                && spans.iter().any(|span| span["syntax"] == "accepted")
            {
                disagreements.push(json!({"diagnostic":id,"kind":"syntax-accepted-compiler-error",
                    "claim":"parser acceptance and compiler rejection concern different verification levels."}));
            }
            rows.push(json!({"id":id,"basis":basis,"core":core}));
        }
        let key = format!("frce-page:{}", hash(&basis)?);
        let (start, end, _) = page(rows.len(), options.limit, options.cursor.as_deref(), &key)?;
        let capture_complete = parsed.cutoffs.is_empty();
        let mut value = json!({"schema":"fr-compiler-evidence-1","revision":self.revision,
            "handle_prefix":format!("frp1:{}:",&self.revision[..32]),"coverage":self.coverage(),
            "basis":basis,"inputs":inputs,"analyzer":analyzer,"mutation_authority":false,
            "budget":{"limit":options.limit,"response_bytes":options.bytes},
            "scope":{"check":options.check,"format":format,"stream":stream,"argv":check["argv"],"cwd":cwd,"covers":check["covers"],
                "toolchain":report["toolchain"],"claim":"declared invocation and retained diagnostic stream; undeclared dependencies remain outside coverage."},
            "outcome":{"passed":check["passed"],"exit_code":check["exit_code"],"execution_complete":execution,"cargo_success":parsed.build_success},
            "capture":{"complete":capture_complete,"cutoffs":parsed.cutoffs,"diagnostics":rows.len(),"errors":errors,
                "ignored_protocol_events":parsed.ignored_events,"retained_bytes":check[stream]["retained_bytes"],"omitted_bytes":check[stream]["omitted_bytes"]},
            "items":rows[start..end],"disagreements":[],"continuation":Value::Null,
            "trust":"trusted caller digest binds retained check output; no independent execution attestation.",
            "claim":"compiler observations and exact syntax locations; no runtime safety or source implementation proof."});
        loop {
            let returned = value["items"].as_array().unwrap().len();
            let end = start + returned;
            let next = (end < rows.len()).then(|| format!("{key}:{end}"));
            value["page"] = json!({"before":start,"returned":returned,"remaining":rows.len()-end,"total":rows.len(),"next":next});
            let ids = value["items"]
                .as_array()
                .unwrap()
                .iter()
                .map(|row| row["id"].clone())
                .collect::<Vec<_>>();
            value["disagreements"] = json!(disagreements
                .iter()
                .filter(|d| ids.contains(&d["diagnostic"]))
                .collect::<Vec<_>>());
            value["disclosure_complete"] = json!(start == 0 && end == rows.len());
            value["complete"] = json!(compiler_evidence_complete(
                execution,
                capture_complete,
                start == 0,
                end == rows.len()
            ));
            value["continuation"] = next.as_ref().map(|cursor| json!({"arguments":["project","compiler-evidence",
                "--from",options.from,"--digest",options.digest,"--check",options.check,"--format",format,
                "--limit",options.limit.to_string(),"--bytes",options.bytes.to_string(),"--cursor",cursor]})).unwrap_or(Value::Null);
            if serde_json::to_vec(&value)?.len() + 512 <= options.bytes {
                break;
            }
            ensure!(
                returned > 1,
                "compiler metadata or one diagnostic exceeds response budget; increase --bytes."
            );
            value["items"].as_array_mut().unwrap().pop();
        }
        Ok(value)
    }

    fn compiler_span(&self, cwd: &str, span: &Value) -> Result<Value> {
        let mut result = json!({"status":"unavailable","reason":"No admitted source location.","occurrence":Value::Null,
            "primary":span["is_primary"].as_bool(),"syntax":"unknown","source":Value::Null});
        if !span["expansion"].is_null() {
            result["status"] = json!("expanded");
            result["reason"] = json!("Macro expansion needs a separate origin relation.");
            return Ok(result);
        }
        let (Some(name), Some(start), Some(end), Some(primary)) = (
            span["file_name"].as_str(),
            span["byte_start"].as_u64(),
            span["byte_end"].as_u64(),
            span["is_primary"].as_bool(),
        ) else {
            return Ok(result);
        };
        let path = Path::new(name);
        let file = if path.is_absolute() {
            path.to_path_buf()
        } else {
            self.root.join(cwd).join(path)
        };
        let Ok(relative) = file.strip_prefix(&self.root) else {
            result["reason"] = json!("Compiler location is outside the selected workspace.");
            return Ok(result);
        };
        if relative
            .components()
            .any(|c| !matches!(c, Component::Normal(_) | Component::CurDir))
        {
            result["reason"] = json!("Compiler location requires unsupported path normalization.");
            return Ok(result);
        }
        let file = self.root.join(
            relative
                .components()
                .filter_map(|c| {
                    if let Component::Normal(part) = c {
                        Some(part)
                    } else {
                        None
                    }
                })
                .collect::<PathBuf>(),
        );
        let Some(source) = self.sources.get(&file) else {
            result["reason"] = json!("Compiler location has no indexed source snapshot.");
            return Ok(result);
        };
        if self
            .index
            .file(&file)
            .is_none_or(|info| info.language != crate::lang::Language::Rust)
        {
            result["reason"] = json!("Compiler location is outside the Rust source subset.");
            return Ok(result);
        }
        if start > end
            || end > source.len() as u64
            || !source.is_char_boundary(start as usize)
            || !source.is_char_boundary(end as usize)
        {
            result["status"] = json!("invalid");
            result["reason"] = json!("Compiler byte span is outside source boundaries.");
            return Ok(result);
        }
        let occurrence = self.occurrence(
            &file,
            crate::span::Span::new(start as usize, end as usize),
            if primary {
                "compiler-primary"
            } else {
                "compiler-secondary"
            },
        )?;
        let mut parses = self.parses.borrow_mut();
        if !parses.contains_key(&file) {
            parses.insert(
                file.clone(),
                crate::parse::Parsers::new().parse(crate::lang::Language::Rust, source)?,
            );
        }
        let parsed = &parses[&file];
        result["status"] = json!("exact");
        result["reason"] = json!("Compiler byte offsets in the validated source snapshot.");
        result["syntax"] = json!(if parsed.has_errors() {
            "rejected"
        } else {
            "accepted"
        });
        result["occurrence"] = json!(occurrence);
        if let Some(node) = self
            .nodes
            .iter()
            .position(|node| node.path == occurrence.path && node.symbol.is_none())
        {
            result["source"] = json!({"arguments":["project","show",self.handle(node),"--source","--bytes","256","--offset",start.to_string()]});
        }
        Ok(result)
    }
}
